src/extension/gogoduck/src/main/java/au/org/emii/gogoduck/worker/GoGoDuck.java

```diff
@@ -0,0 +1,511 @@
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.opengis.util.ProgressListener;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.zip.GZIPInputStream;
+
+public class GoGoDuck {
+    private static final Logger logger = LoggerFactory.getLogger(GoGoDuck.class);
+
+    private static final Map<String, String> replacePrefixes = new HashMap<String, String>();
+    static {
+        replacePrefixes.put("/mnt/imos-t3/", "https://data.aodn.org.au/");
+        replacePrefixes.put("/mnt/opendap/2/SRS/", "https://thredds.aodn.org.au/thredds/fileServer/srs/");
+    }
+
+    public static final String ncksPath = "/usr/bin/ncks";
+    public static final String ncrcatPath = "/usr/bin/ncrcat";
+    public static final String ncdpqPath = "/usr/bin/ncpdq";
+
+    private final String geoserver;
+    private final String profile;
+    private final String subset;
+    private final Path outputFile;
+    private final Integer limit;
+    private Path baseTmpDir;
+    private int threadCount = 1;
+    private UserLog userLog = null;
+    private ProgressListener progressListener = null;
+
+    public GoGoDuck(String geoserver, String profile, String subset, String outputFile, Integer limit) {
+        this.geoserver = geoserver;
+        this.profile = profile;
+        this.subset = subset;
+        this.outputFile = new File(outputFile).toPath();
+        this.limit = limit;
+        this.baseTmpDir = new File(System.getProperty("java.io.tmpdir")).toPath();
+        this.userLog = new UserLog();
+    }
+
+    public void setTmpDir(String tmpDir) {
+        this.baseTmpDir = new File(tmpDir).toPath();
+    }
+
+    public void setThreadCount(int threadCount) {
+        logger.info(String.format("Setting thread count to %d", threadCount));
+        this.threadCount = threadCount;
+    }
+
+    public void setUserLog(String userLog) {
+        this.userLog = new UserLog(userLog);
+    }
+
+    public void setProgressListener(ProgressListener progressListener) {
+        this.progressListener = progressListener;
+    }
+
+    public boolean isCancelled() {
+        if (null != progressListener) {
+            return progressListener.isCanceled();
+        }
+        else {
+            return false;
+        }
+    }
+
+    private void throwIfCancelled() throws GoGoDuckException {
+        if (isCancelled())
+            throw new GoGoDuckException("Job cancelled");
+    }
+
+    public void run() {
+        GoGoDuckModule module = getProfileModule(profile, geoserver, subset, userLog);
+
+        Path tmpDir = null;
+
+        try {
+            tmpDir = Files.createTempDirectory(baseTmpDir, "gogoduck");
+
+            URIList URIList = module.getUriList();
+
+            enforceFileLimit(URIList, limit);
```

Comment:
This is obviously the core high level algorithm, and I think it gives some clues as to how I *think* this (IMO) big class could be refactored.  I know we have differences of opinion in this area, but *IMO* adding some level of abstraction makes it easier for maintainers (e.g. you usually have less code to understand when making a change, as a more specific class is smaller).

I would guess at least specific classes for:

* downloading files (including enforcement of file limit)
* subsetting (+ post process)
* aggregating (+ update metadata)

would make sense.